### Importamos librerias

In [ ]:
# Tratamiento de datos
import pandas as pd

# Exploración de archivos
import sys
import os

# Funciones personalizadas
sys.path.append(os.path.abspath('../src'))
import sp_eda as se

### Para mostrar todas las columnas

In [26]:
pd.set_option('display.max_columns', None)

### Cargamos los conjuntos de datos desde un archivo CSV

In [27]:
df_sales = pd.read_csv('../data/raw/sales.csv')
df_customers = pd.read_csv('../data/raw/customers.csv')
df_products = pd.read_csv('../data/raw/products.csv')
df_categories = pd.read_csv('../data/raw/categories.csv')
df_cities = pd.read_csv('../data/raw/cities.csv')
df_countries = pd.read_csv('../data/raw/countries.csv')

### Información sobre el conjunto de datos

La base de datos de ventas de comestibles es un conjunto de datos de 6 archivos:

1. **sales**:  Contiene datos transaccionales de cada venta.
 - SalesID:  Identificador único de cada venta.
 - SalesPersonID:  Empleado responsable de la venta.
 - CustomerID:  Cliente que realiza la compra.
 - ProductID:  Producto en venta.
 - Quantity:  Número de unidades vendidas.
 - Discount:  Descuento aplicado a la venta.
 - TotalPrice:  Precio final de venta después de descuentos. 
 - SalesDate:  Fecha y hora de la venta.
 - TransactionNumber:  Identificador único de la transacción.

2. **customers**:  Contiene información sobre los clientes que realizan compras.
 - CustomerID:  Identificador único para cada cliente.
 - FirstName:  Nombre del cliente.
 - MiddleInitial:  Inicial media del cliente.
 - LastName:  Apellido del cliente.
 - cityID:  Ciudad del cliente.
 - Address:  Dirección residencial del cliente.

4. **products**:  Almacena detalles sobre los productos que se venden.
 - ProductID:  Identificador único para cada producto.
 - ProductName:  Nombre del producto.
 - Price:  Precio por unidad del producto.
 - CategoryID:  Identificador de categoría único
 - Class:  Clasificación del producto.
 - ModifyDate:  Fecha de última modificación.
 - Resistant:  Categoría de resistencia del producto.
 - IsAllergic:  Indica si el artículo es un alérgeno
 - VitalityDays:  Clasificación de tipo vital de producto.

5. **categories**:  Define las categorías de los productos.
 - CategoryID:  Identificador único para cada categoría de producto.
 - CategoryName:  Nombre de la categoría de producto.

6. **cities**:  Contiene datos geográficos a nivel de ciudad.
 - CityID:  Identificador único para cada ciudad.
 - CityName:  Nombre de la ciudad.
 - Zipcode:  Población de la ciudad.
 - CountryID:  Referencia al país correspondiente.

7. **countries**:  Almacena metadatos relacionados con el país.
 - CountryID:  Identificador único para cada país.
 - CountryName:  Nombre del país.
 - CountryCode:  Código de país de dos letras.

### Análisis preliminar

In [28]:
se.analisis_rapido(df_sales)

Las 5 primeras columnas de Dataframe son:


,SalesID,SalesPersonID,CustomerID,ProductID,Quantity,Discount,TotalPrice,SalesDate,TransactionNumber
0,1,6,27039,381,7,0.0,0.0,2018-02-05 07:38:25.430,FQL4S94E4ME1EZFTG42G
1,2,16,25011,61,7,0.0,0.0,2018-02-02 16:03:31.150,12UGLX40DJ1A5DTFBHB8
2,3,13,94024,23,24,0.0,0.0,2018-05-03 19:31:56.880,5DT8RCPL87KI5EORO7B0
3,4,8,73966,176,19,0.2,0.0,2018-04-07 14:43:55.420,R3DR9MLD5NR76VO17ULE
4,5,10,32653,310,9,0.0,0.0,2018-02-12 15:37:03.940,4BGS0Z5OMAZ8NDAFHHP3


Información básica de Dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6758125 entries, 0 to 6758124
Data columns (total 9 columns):
 #   Column             Dtype  
---  ------             -----  
 0   SalesID            int64  
 1   SalesPersonID      int64  
 2   CustomerID         int64  
 3   ProductID          int64  
 4   Quantity           int64  
 5   Discount           float64
 6   TotalPrice         float64
 7   SalesDate          object 
 8   TransactionNumber  object 
dtypes: float64(2), int64(5), object(2)
memory usage: 464.0+ MB


None

El número de duplicados es: 0
El número de valores nulos es:


SalesID                  0
SalesPersonID            0
CustomerID               0
ProductID                0
Quantity                 0
Discount                 0
TotalPrice               0
SalesDate            67526
TransactionNumber        0
dtype: int64

In [29]:
se.analisis_rapido(df_customers)

Las 5 primeras columnas de Dataframe son:


,CustomerID,FirstName,MiddleInitial,LastName,CityID,Address
0,1,Stefanie,Y,Frye,79,97 Oak Avenue
1,2,Sandy,T,Kirby,96,52 White First Freeway
2,3,Lee,T,Zhang,55,921 White Fabien Avenue
3,4,Regina,S,Avery,40,75 Old Avenue
4,5,Daniel,S,Mccann,2,283 South Green Hague Avenue


Información básica de Dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 98759 entries, 0 to 98758
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   CustomerID     98759 non-null  int64 
 1   FirstName      98759 non-null  object
 2   MiddleInitial  97782 non-null  object
 3   LastName       98759 non-null  object
 4   CityID         98759 non-null  int64 
 5   Address        98759 non-null  object
dtypes: int64(2), object(4)
memory usage: 4.5+ MB


None

El número de duplicados es: 0
El número de valores nulos es:


CustomerID         0
FirstName          0
MiddleInitial    977
LastName           0
CityID             0
Address            0
dtype: int64

In [30]:
se.analisis_rapido(df_products)

Las 5 primeras columnas de Dataframe son:


,ProductID,ProductName,Price,CategoryID,Class,ModifyDate,Resistant,IsAllergic,VitalityDays
0,1,Flour - Whole Wheat,74.2988,3,Medium,2018-02-16 08:21:49.190,Durable,Unknown,0.0
1,2,Cookie Chocolate Chip With,91.2329,3,Medium,2017-02-12 11:39:10.970,Unknown,Unknown,0.0
2,3,Onions - Cippolini,9.1379,9,Medium,2018-03-15 08:11:51.560,Weak,False,111.0
3,4,"Sauce - Gravy, Au Jus, Mix",54.3055,9,Medium,2017-07-16 00:46:28.880,Durable,Unknown,0.0
4,5,Artichokes - Jerusalem,65.4771,2,Low,2017-08-16 14:13:35.430,Durable,True,27.0


Información básica de Dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ProductID     452 non-null    int64  
 1   ProductName   452 non-null    object 
 2   Price         452 non-null    float64
 3   CategoryID    452 non-null    int64  
 4   Class         452 non-null    object 
 5   ModifyDate    452 non-null    object 
 6   Resistant     452 non-null    object 
 7   IsAllergic    452 non-null    object 
 8   VitalityDays  452 non-null    float64
dtypes: float64(2), int64(2), object(5)
memory usage: 31.9+ KB


None

El número de duplicados es: 0
El número de valores nulos es:


ProductID       0
ProductName     0
Price           0
CategoryID      0
Class           0
ModifyDate      0
Resistant       0
IsAllergic      0
VitalityDays    0
dtype: int64

In [31]:
se.analisis_rapido(df_categories)

Las 5 primeras columnas de Dataframe son:


,CategoryID,CategoryName
0,1,Confections
1,2,Shell fish
2,3,Cereals
3,4,Dairy
4,5,Beverages


Información básica de Dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   CategoryID    11 non-null     int64 
 1   CategoryName  11 non-null     object
dtypes: int64(1), object(1)
memory usage: 308.0+ bytes


None

El número de duplicados es: 0
El número de valores nulos es:


CategoryID      0
CategoryName    0
dtype: int64

In [32]:
se.analisis_rapido(df_cities)

Las 5 primeras columnas de Dataframe son:


,CityID,CityName,Zipcode,CountryID
0,1,Dayton,80563,32
1,2,Buffalo,17420,32
2,3,Chicago,44751,32
3,4,Fremont,20641,32
4,5,Virginia Beach,62389,32


Información básica de Dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   CityID     96 non-null     int64 
 1   CityName   96 non-null     object
 2   Zipcode    96 non-null     int64 
 3   CountryID  96 non-null     int64 
dtypes: int64(3), object(1)
memory usage: 3.1+ KB


None

El número de duplicados es: 0
El número de valores nulos es:


CityID       0
CityName     0
Zipcode      0
CountryID    0
dtype: int64

In [33]:
se.analisis_rapido(df_countries)

Las 5 primeras columnas de Dataframe son:


,CountryID,CountryName,CountryCode
0,1,Armenia,AN
1,2,Canada,FO
2,3,Belize,MK
3,4,Uganda,LV
4,5,Thailand,VI


Información básica de Dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CountryID    206 non-null    int64 
 1   CountryName  206 non-null    object
 2   CountryCode  205 non-null    object
dtypes: int64(1), object(2)
memory usage: 5.0+ KB


None

El número de duplicados es: 0
El número de valores nulos es:


CountryID      0
CountryName    0
CountryCode    1
dtype: int64

### Observaciones:

Los datos crudos incluyen tanto variables numéricas (int, float) como variables categóricas (object). 

No hay valores duplicados. 

Hay valores nulos en los archivos:

df_sales: columna 'SalesDate' tiene 67526 nulos

df_customers: columna 'MiddleInitial' tiene 977 nulos

df_countries: columna 'CountryCode' tiene 1 nulo

Columna 'TotalPrice' de df_sales tiene valor unico: 0.00. 
Esto significa que no se ha calculado correctamente y es necesario calcularlo.


In [34]:
df_sales['TotalPrice'].describe()

count    6758125.0
mean           0.0
std            0.0
min            0.0
25%            0.0
50%            0.0
75%            0.0
max            0.0
Name: TotalPrice, dtype: float64

In [35]:
# Comprobamos las dimensiones de cada archivo

print(f"df_sales: {df_sales.shape}")
print(f"df_customers: {df_customers.shape}")
print(f"df_products: {df_products.shape}")
print(f"df_categories: {df_categories.shape}")
print(f"df_cities: {df_cities.shape}")
print(f"df_countries: {df_countries.shape}")

df_sales: (6758125, 9)
df_customers: (98759, 6)
df_products: (452, 9)
df_categories: (11, 2)
df_cities: (96, 4)
df_countries: (206, 3)
